### ISSUE TITLE
[web] Can't add RN modules with unusual names that need Babeling

### ISSUE THREAD

**Issue by sfcgeorge** [expo/expo-cli#556]
**Update** packages prefixed `react-native` work, only other namings don't. See https://github.com/expo/expo-cli/issues/556#issuecomment-490014963

## Description

A simple React Native module like this https://github.com/hectahertz/react-native-typography has no native code and only uses RN APIs that RNW supports (Platform). So it should be possible to add to an Expo Web project.

## Expected Behavior

Should be able to yarn add react-native-typography (or similar modules), possibly add to some config somewhere, and have it work in Expo Web as well as native. Perhaps there could be a new key in App.js or something:

```js
{
  rnw_compatible_modules: [
    "react-native-typography"
  ]
]
```

## Observed Behavior

Depending on the module the build fails complaining about undefined things from RN (like `Platform`) or ES6 that hasn't been Babeled.

## Sleuth

I see @EvanBacon has done a few [PRs](https://github.com/expo/expo-cli/pull/535) on the auto webpack and babel config and some customisation is possible, but I can't see how to add to the list of paths Babel will Babel.

In [createBabelLoader](https://github.com/expo/expo-cli/blob/master/packages/webpack-config/webpack/loaders/createBabelLoader.js#L33) there is an argument to pass in pathsToInclude which could be intended for this purpose. 

But [where it's called](https://github.com/expo/expo-cli/blob/3078a4a110b5b6c1540f071ee6dfb9f1f1947d95/packages/webpack-config/webpack/webpack.config.unimodules.js#L76) there's no way to pass in the extra paths / module names.

PS: other than this Expo Web has been working great for me on a new app, it's AWESOME! Well done @EvanBacon et al, I can't wait for SDK 33 to use on native too 😄

**Comment by ide** [comment]
Importing Platform should work since it's part of RNW's public API. It sounds like the issue is related to Babel -- in addition to your observation that JS isn't being compiled down, Babel is also responsible for rewriting react-native imports to react-native-web imports, which could be causing the Platform bug as well.

**Comment by sfcgeorge** [comment]
Yep, for example `react-navigation` doesn't come with Expo and uses RN features, but it works because it's been added to the list Babel does stuff to here https://github.com/expo/expo-cli/blob/3078a4a110b5b6c1540f071ee6dfb9f1f1947d95/packages/webpack-config/webpack/loaders/createBabelLoader.js#L9

So I think there needs to be a simple way to add other packages to that list, to maintain the excellent Expo config simplicity :)

**Comment by EvanBacon** [comment]
This doesn't seem correct. Could you please provide a repro. This line should include any module that starts with `react-native` including `react-native-typography`: https://github.com/expo/expo-cli/blob/3078a4a110b5b6c1540f071ee6dfb9f1f1947d95/packages/webpack-config/webpack/loaders/createBabelLoader.js#L8

**Comment by chrisdrackett** [comment]
@EvanBacon I think this issue is about being able to add whatever modules we want to this list vs. just adding `react-native-typography`. For example I need to add a local module named `calm-components` to this list (using a fork currently)

**Comment by sfcgeorge** [comment]
Oh thanks for pointing that out @EvanBacon, I assumed it would need a glob wildcard to work. I've just tried some more RN packages and they work great! 🎉  Working 3rd party RN modules:

* react-native-calendars
* react-native-segmented-control-tab

This is really neat Evan, for a large number of RN packages people can just `yarn add` and not even think about Babel or Webpack. Love it ❤️ 

---

One that didn't work is `rn-sliding-up-panel` — now it's clear that the name needs to start with `react-native` to be picked up. Perhaps you could add `rn-` to that prefix list, I don't know if that would cause any false positives. It is an example of @chrisdrackett's point that there are modules which don't fit the pattern so it would still be nice to have an easy way to add their paths, but it seems less urgent now as local modules could easily be prefixed and hopefully most popular 3rd party ones already follow naming conventions.

Next, as the prefix is correct I don't know why `react-native-typography` isn't working; I get the error "Cannot read property 'Platform' of undefined" on this line:

```js
letterSpacing: _reactNative.Platform.OS === "ios" ? (0, _sanFranciscoSpacing2.default)(34) : undefined,
```

Their docs say it supports RNW which is why I assumed this was an Expo issue. Not sure where that `_reactNative` is coming from but I assume that's preventing the webpack alias from being picked up. I'll open an issue over on that repo.

---

I've renamed the issue slightly and linked to this comment. Up to y'all whether you want to keep it open for the custom names aspect, open a new clearer issue, or close for now if you want to encourage the naming convention. 

---

### TLDR

* ✅ Packages prefixed `react-native` get Babeled perfectly
* 🔶 Packages with other prefixes aren't picked up yet
* 🤷‍♀ Some packages need testing and fixing to work with RNW but it's not an Expo problem

**Comment by EvanBacon** [comment]
I think ideally you should be able to extend the webpack config to add the babel alias as this would cuz the least amount of confusion. At the moment the expo/webpack-config is changing a lot so I wouldn't recommend reaching into the internals just yet. If you create a `webpack.config.js` in your root folder, then expo-cli will use this in place of the default.

For the time being I've added a field to the `app.json` [`expo.web.build.babel.include: []`](https://github.com/expo/expo-cli/blob/7c847762fff62913232b0dba95b18a376e2d9d06/packages/webpack-config/webpack/webpack.common.js#L241) which you can [pass library names to](https://github.com/expo/expo-cli/blob/7c847762fff62913232b0dba95b18a376e2d9d06/packages/webpack-config/webpack/loaders/createBabelLoader.js#L37-L49). 

`app.json`
```js
"expo": {
  "web": {
    "build": {
      "include": ["rn-sliding-up-panel"]
    }
  }
}
```

> 🤷‍♀ Some packages need testing and fixing to work with RNW but it's not an Expo problem

I just want it to work 😃 if there's a library that doesn't work please feel free to link me, tweet me (baconbrix) or open an issue.

**Comment by sfcgeorge** [comment]
This is really great thank you, adding modules to the include array works perfectly. I had to make a small fix to the module then it worked. So cool that even something complicated like sliding panel works seemingly perfectly on the web 👍
